### ISSUE TITLE
12-nlp-url-spam.ipynb

### ISSUE THREAD

**Issue by mezcolantriz** [4GeeksAcademy/Antonio27M-machine-learning#7]
## LO QUE HAS HECHO BIEN

- **Importaciones completas**: pandas, regex, seaborn, matplotlib, pickle, nltk, WordCloud, TfidfVectorizer, train_test_split, GridSearchCV, métricas, SVC, Pipeline
- **Limpieza de datos correcta**: eliminación de duplicados y reset de índice
- **Conversión del target**: `is_spam` a 0/1 correctamente
- **Preprocesamiento de texto completo**:
  - Lowercase
  - Eliminación de caracteres especiales
  - Tokenización
  - Lematización con WordNetLemmatizer
  - Eliminación de stopwords
  - Filtrado por longitud de palabra (>3)
- **Visualización con WordCloud** 
- **TfidfVectorizer bien configurado**: `max_features=5000`, `max_df=0.8`, `min_df=5`
- **Train/test split ANTES de entrenar** 
- **Modelo SVC con kernel lineal** (apropiado para texto)
- **Métricas completas**: accuracy, recall, f1, precision
- **Hiperparametrización con GridSearchCV**: probando C, kernel, gamma
- **Pipeline con StandardScaler** (aunque no es necesario para SVC con kernel lineal en texto TF-IDF)
- **Matriz de confusión visualizada**
- **Modelo guardado con pickle** 

### 1. **Falta guardar el vectorizer (IMPORTANTE)**
El modelo necesita el vectorizer para hacer predicciones con datos nuevos:
```python
# Solo guardas el modelo
with open('modelo-npl-url().pkl', 'wb') as file:
    pickle.dump(grid_svc.best_estimator_, file)

# Deberías guardar también el vectorizer
with open('vectorizer_url_spam.pkl', 'wb') as file:
    pickle.dump(vectorizer, file)
```

### 2. **Matriz de confusión usa predicciones del modelo base, no del optimizado**
```python
# Usas y_pred (del modelo sin optimizar)
df_confusion = confusion_matrix(y_test, y_pred)

# Deberías usar y_pred_dos (del mejor modelo)
df_confusion = confusion_matrix(y_test, y_pred_dos)
```
### 3. **¿Conclusiones finales? ¿Interpretación de resultados? :( **
---

Muy buen trabajo @Antonio27M . Está bastante completo y correctamente ejecutado a excepción de un par de cosillas. Los 3 puntos principales a mejorar son guardar el vectorizer junto con el modelo para predicciones nuevas, usar el modelo optimizado para la matriz de confusión y las conclusiones y observaciones.
